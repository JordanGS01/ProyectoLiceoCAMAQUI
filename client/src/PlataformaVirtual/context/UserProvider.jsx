

import { useState } from 'react'
import { UserContext } from './UserContext'


const userDataInitialState = {
    nombre: '',
    cedula: 0,
    tipo: ''
}

export const UserProvider = ({ children }) => {

    const [userData, setUserData] = useState(userDataInitialState);
    const [loged, setLoged] = useState(false);

    const setUser = ({ nombre, cedula, tipo }) => {
        setUserData({
            nombre,
            cedula,
            tipo
        })
        setLoged( true );
    }

    const logOutUser = () => {
        setUserData(userDataInitialState);
        setLoged( false );
    }

    return (
        <UserContext.Provider value={{ 
            userData,
            setUser,
            logOutUser,
            loged
        }}>
            { children }
        </UserContext.Provider>
    )
}