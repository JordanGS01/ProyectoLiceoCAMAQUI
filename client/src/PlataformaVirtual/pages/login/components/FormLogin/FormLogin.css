

.FormLogin-ContenedorBotones{
    display: flex;
    margin-left: auto;
}

.FormLogin-Form{
    display: flex;
    flex-direction: column;
}