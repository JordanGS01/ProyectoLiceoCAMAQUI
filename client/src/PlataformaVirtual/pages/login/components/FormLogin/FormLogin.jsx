

import { useState, useContext, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';

import { Box, FormControl, TextField, Button, InputAdornment, IconButton, Typography } from '@mui/material'
import { BadgeOutlined, PasswordOutlined, Visibility, VisibilityOff } from '@mui/icons-material'

import { UserContext } from '../../../../context/UserContext';
import { Alert } from '../../../../../ui/components/Alert';

import './FormLogin.css'
import { useForm } from '../../../../hooks';
import { onFormSubmit } from '../../helpers';


export const FormLogin = () => {
    const { setUser, loged, isStudent, isProfessor, isAdmin  } = useContext( UserContext );
    const [showPassword, setShowPassword] = useState(false);
    const [openAlert, setOpenAlert] = useState(false);

    const navigate = useNavigate();

    const { formState, onInputChange } = useForm({
        cedula: '',
        contra: ''
    });
    const { cedula, contra } = formState;
    

    useEffect(() => {
      if( loged && (isStudent() || isProfessor()) ) {
        navigate('/menuInicial');
      } else if (loged && isAdmin()) {
        navigate('/admin');
      }
    }, [loged])
    
    return (
    <Box
        component="span" 
        width='80vh'
        sx={{   p: 2,
                border: '1px solid grey',
                background: '#4FA4D3',
                borderRadius: '20px',
                display: 'flex',
                flexDirection: 'column',
            }}
    >
        <form className='FormLogin-Form' onSubmit={ (e) => { onFormSubmit(e, formState, setUser, setOpenAlert) } }>

            <Typography variant="h3" gutterBottom sx={{ ml: 2, color: '#FFFFFF' }}>
                Iniciar sesión
            </Typography>
            
            {/* Sección de la cédula */}
            <FormControl sx={{ m: 2 }}>
                <TextField
                    required
                    autoFocus
                    label='Número de cédula'
                    id='cedula-usuario'
                    InputProps={{
                        startAdornment:(
                            <InputAdornment position='start'>
                                <BadgeOutlined />
                            </InputAdornment>
                        )
                    }}
                    sx={{ backgroundColor: '#FFFDFD', borderRadius: '5px' }}
                    variant='filled'
                    type='number'
                    aria-label='cedula'

                    value={ cedula }
                    name='cedula'
                    onChange={ onInputChange }
                />
            </FormControl>

            {/* Sección de la contraseña */}
            <FormControl sx={{ m: 2 }}>
                <TextField
                    required
                    label='Contraseña'
                    id='contra-usuario'
                    type={ showPassword ? 'text' : 'password' }
                    InputProps={{
                        startAdornment:(
                            <InputAdornment position='start'>
                                <PasswordOutlined />
                            </InputAdornment>
                        ),
                        endAdornment:(
                            <InputAdornment position='end'>
                                <IconButton
                                    aria-label='toggle password visibility'
                                    onClick={ () => setShowPassword(!showPassword) }
                                    onMouseDown={ (event) => event.preventDefault() }
                                >
                                    { showPassword ? <VisibilityOff /> : <Visibility /> }
                                </IconButton>
                            </InputAdornment>
                        )
                    }}
                    sx={{ backgroundColor: '#FFFDFD', borderRadius: '5px' }}
                    variant='filled'
                    aria-label='contra'

                    value={ contra }
                    name='contra'
                    onChange={ onInputChange }
                />

            </FormControl>

            {/* Sección de recuperar contraseña */}
            <Typography variant="caption" gutterBottom sx={{ ml: 'auto', mr: 2, mb: 2, color: '#FFFFFF' }}>
                ¿Olvidó su contraseña? Haga click aquí
            </Typography>

            {/* Sección de la botones */}
            <div className='FormLogin-ContenedorBotones'>

                <Button
                    variant="contained"
                    onClick={ () => { navigate('/signin') } }
                >
                    Registrarse
                </Button>

                <Button
                    variant="contained"
                    sx={{ ml: 1, mr: 2 }}
                    type='submit'
                >
                    Ingresar
                </Button>

            </div>

        </form>

        {/* Alert que se mostrará cuando los datos ingresados no estén registrados*/}
        <Alert 
            open = { openAlert }
            handleClose = {() => { setOpenAlert(false) }}
            title = "Usuario no encontrado"
            content = "El número de cédula o la contraseña ingresada son erroneas"
            acceptButtonText = "Cerrar"
            acceptButtonFunction = {() => { setOpenAlert(false) }}
            oneButton = { true }
        />

    </Box>
  )
}
