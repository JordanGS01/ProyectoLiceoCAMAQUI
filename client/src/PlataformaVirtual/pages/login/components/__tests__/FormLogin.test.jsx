

import { cleanup, render, screen, fireEvent } from '@testing-library/react'

import { BrowserRouter } from 'react-router-dom' 

import { UserContext } from '../../../../context/UserContext'

import { FormLogin } from '../FormLogin/FormLogin'


describe('Pruebas sobre el componente <FormLogin />', () => { 
    
    let containerFormLogin;

    beforeEach(() => {
        const userData = {};
        const setUser = () => {};
        const logOutUser = () => {};
        const loged = false;
        const isStudent = () => {};
        const isProfessor = () => {};
        const isAdmin = () => {};

        const { container } = render(
            <BrowserRouter>
                <UserContext.Provider value={{ 
                    userData, setUser, logOutUser,
                    loged, isStudent, isProfessor,
                    isAdmin
                }}>
                    <FormLogin />
                </UserContext.Provider>
            </BrowserRouter>
        );

        containerFormLogin = container;
    })

    
    test('Debe hacer match con el snapshot', () => {
        expect( containerFormLogin ).toMatchSnapshot();
    })

    test('El valor del input cedula debe cambiar', () => { 
        const cedula = 208140251;

        const inputCedula = screen.getByRole('spinbutton', { name: 'Número de cédula' });
        expect(inputCedula.value).toEqual('');
        fireEvent.input(inputCedula, { target: { value: cedula } });
        expect(inputCedula.value).toEqual(`${cedula}`);
    })

    test('El valor del input contraseña debe cambiar', () => { 
        const contra = "contraSegura";

        const botonVisibilidad = screen.getByRole('button', { name: 'toggle password visibility' });
        fireEvent.click(botonVisibilidad);
        
        const inputContra = screen.getByRole('textbox', { name: 'Contraseña' });
        expect(inputContra.value).toEqual('');
        fireEvent.input(inputContra, { target: { value: contra } });
        expect(inputContra.value).toEqual(`${contra}`);
    })

    test('La contraseña debe ser visible al dar click en el botón de visibilidad', () => { 
        const botonVisibilidad = screen.getByRole('button', { name: 'toggle password visibility' });
        
        fireEvent.click(botonVisibilidad);
        expect(screen.getByRole('textbox', { name: 'Contraseña' })).toBeTruthy();
    })

    test('El botón "Registrarse" redirecciona al SignIn', () => { 
        const botonRegistrarse = screen.getByRole('button', { name: 'Registrarse' });

        expect(fireEvent.click(botonRegistrarse)).toBeTruthy();
    })


    afterEach(() => {
        cleanup();
    })

})