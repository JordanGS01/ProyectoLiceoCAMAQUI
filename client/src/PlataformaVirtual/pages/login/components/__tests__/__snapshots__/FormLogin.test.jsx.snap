// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre el componente <FormLogin /> Debe hacer match con el snapshot 1`] = `
<div>
  <span
    class="MuiBox-root css-1ytefdo"
  >
    <form
      class="FormLogin-Form"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom css-rlscyn-MuiTypography-root"
      >
        Iniciar sesión
      </h3>
      <div
        class="MuiFormControl-root css-k2e4wk-MuiFormControl-root"
      >
        <div
          aria-label="cedula"
          class="MuiFormControl-root MuiTextField-root css-mcjkg0-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-colorPrimary Mui-focused Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled css-o943dk-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="cedula-usuario"
            id="cedula-usuario-label"
          >
            Número de cédula
            <span
              aria-hidden="true"
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl MuiInputBase-adornedStart css-1xab92c-MuiInputBase-root-MuiFilledInput-root"
          >
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-filled MuiInputAdornment-sizeMedium css-9mc7sf-MuiInputAdornment-root"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="BadgeOutlinedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M14 12h4v1.5h-4zm0 3h4v1.5h-4z"
                />
                <path
                  d="M20 7h-5V4c0-1.1-.9-2-2-2h-2c-1.1 0-2 .9-2 2v3H4c-1.1 0-2 .9-2 2v11c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V9c0-1.1-.9-2-2-2zm-9 0V4h2v5h-2V7zm9 13H4V9h5c0 1.1.9 2 2 2h2c1.1 0 2-.9 2-2h5v11z"
                />
                <circle
                  cx="9"
                  cy="13.5"
                  r="1.5"
                />
                <path
                  d="M11.08 16.18c-.64-.28-1.34-.43-2.08-.43s-1.44.15-2.08.43c-.56.24-.92.78-.92 1.39V18h6v-.43c0-.61-.36-1.15-.92-1.39z"
                />
              </svg>
            </div>
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputAdornedStart css-1vbc0rj-MuiInputBase-input-MuiFilledInput-input"
              id="cedula-usuario"
              name="cedula"
              required=""
              type="number"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="MuiFormControl-root css-k2e4wk-MuiFormControl-root"
      >
        <div
          aria-label="contra"
          class="MuiFormControl-root MuiTextField-root css-mcjkg0-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled css-o943dk-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="contra-usuario"
            id="contra-usuario-label"
          >
            Contraseña
            <span
              aria-hidden="true"
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd css-1cxkmns-MuiInputBase-root-MuiFilledInput-root"
          >
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-filled MuiInputAdornment-sizeMedium css-9mc7sf-MuiInputAdornment-root"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="PasswordOutlinedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M2 17h20v2H2v-2zm1.15-4.05L4 11.47l.85 1.48 1.3-.75-.85-1.48H7v-1.5H5.3l.85-1.47L4.85 7 4 8.47 3.15 7l-1.3.75.85 1.47H1v1.5h1.7l-.85 1.48 1.3.75zm6.7-.75 1.3.75.85-1.48.85 1.48 1.3-.75-.85-1.48H15v-1.5h-1.7l.85-1.47-1.3-.75L12 8.47 11.15 7l-1.3.75.85 1.47H9v1.5h1.7l-.85 1.48zM23 9.22h-1.7l.85-1.47-1.3-.75L20 8.47 19.15 7l-1.3.75.85 1.47H17v1.5h1.7l-.85 1.48 1.3.75.85-1.48.85 1.48 1.3-.75-.85-1.48H23v-1.5z"
                />
              </svg>
            </div>
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd css-1o2n4nu-MuiInputBase-input-MuiFilledInput-input"
              id="contra-usuario"
              name="contra"
              required=""
              type="password"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-filled MuiInputAdornment-sizeMedium css-jv61v3-MuiInputAdornment-root"
            >
              <button
                aria-label="toggle password visibility"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="VisibilityIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <span
        class="MuiTypography-root MuiTypography-caption MuiTypography-gutterBottom css-mzb9o7-MuiTypography-root"
      >
        ¿Olvidó su contraseña? Haga click aquí
      </span>
      <div
        class="FormLogin-ContenedorBotones"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Registrarse
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-wdj7wd-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Ingresar
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </form>
  </span>
</div>
`;
